src/thread_manager.ts

```diff
@@ -56,279 +61,227 @@ export class ThreadManager<SCG extends ExtendableGenerics = DefaultGenerics> {
-    // re-enable functionality bellow, and take into consideration sorting
-
-    // re-adjust next/previous cursors based on query options
-    // data.threads.forEach((thread) => {
-    //   thread.state.next((current) => ({
-    //     ...current,
-    //     nextCursor: current.latestReplies.length < optionsWithDefaults.reply_limit ? null : current.nextCursor,
-    //     previousCursor:
-    //       current.latestReplies.length < optionsWithDefaults.reply_limit ? null : current.previousCursor,
-    //   }));
-    // });
-
-    return { threads, next };
+  public queryThreads = (options: QueryThreadsOptions = {}) => {
+    return this.client.queryThreads({
+      limit: 25,
+      participant_limit: 10,
+      reply_limit: 10,
+      watch: true,
+      ...options,
+    });
   };
 
-  // remove `next` from options as that is handled internally
   public loadNextPage = async (options: Omit<QueryThreadsOptions, 'next'> = {}) => {
-    const { nextCursor, loadingNextPage } = this.state.getLatestValue();
-
-    if (nextCursor === null || loadingNextPage) return;
+    const { pagination } = this.state.getLatestValue();
 
-    const optionsWithNextCursor: QueryThreadsOptions = {
-      ...options,
-      next: nextCursor,
-    };
-
-    this.state.next((current) => ({ ...current, loadingNextPage: true }));
+    if (pagination.isLoadingNext || !pagination.nextCursor) return;
 
     try {
-      const data = await this.queryThreads(optionsWithNextCursor);
+      this.state.partialNext({ pagination: { ...pagination, isLoadingNext: true } });
+
+      const response = await this.queryThreads({
+        ...options,
+        next: pagination.nextCursor,
+      });
 
       this.state.next((current) => ({
         ...current,
-        threads: data.threads.length ? current.threads.concat(data.threads) : current.threads,
-        nextCursor: data.next ?? null,
+        ...prepareThreadsUpdate(
+          current,
+          response.threads.length ? current.threads.concat(response.threads) : current.threads,
+        ),
+        pagination: {
+          ...current.pagination,
+          nextCursor: response.next ?? null,
+          isLoadingNext: false,
+        },
       }));
     } catch (error) {
       this.client.logger('error', (error as Error).message);
-    } finally {
-      this.state.next((current) => ({ ...current, loadingNextPage: false }));
+      this.state.next((current) => ({
+        ...current,
+        pagination: {
+          ...current.pagination,
+          isLoadingNext: false,
+        },
+      }));
     }
   };
 }
+
+function prepareThreadsUpdate<SCG extends ExtendableGenerics = DefaultGenerics>(
+  state: ThreadManagerState<SCG>,
+  threads: Thread<SCG>[],
+): Partial<ThreadManagerState<SCG>> {
+  if (threads === state.threads) {
+    return {};
+  }
+
+  return {
+    threads,
+    threadsById: threads.reduce<Record<string, Thread<SCG>>>((acc, thread) => {
```

Comment:
nit:
```suggestion
    threadsById: threads.reduce<Record<string, Thread<SCG>>>((newThreadsById, thread) => {
```

Reply:
Fixed in https://github.com/GetStream/stream-chat-js/pull/1348/commits/07e588099c83a7fa2f24440a9bfe3ebb5a848c8a :)